File: mavlink_raw_comm/src/mavlink_encoder/MavlinkEncoder.cpp
/*********************************************************************
*  Software License Agreement (BSD License)
*
*   Copyright (c) 2022
*   All rights reserved.
*
*   Redistribution and use in source and binary forms, with or without
*   modification, are permitted provided that the following conditions
*   are met:
*    * Redistributions of source code must retain the above copyright
*      notice, this list of conditions and the following disclaimer.
*    * Redistributions in binary form must reproduce the above
*      copyright notice, this list of conditions and the following
*      disclaimer in the documentation and/or other materials provided
*      with the distribution.
*    * Neither the name of the author nor the names of its
*      contributors may be used to endorse or promote products derived
*      from this software without specific prior written permission.
*   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
*   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
*   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
*   FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
*   COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
*   INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
*   BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
*   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
*   CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
*   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
*   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*   POSSIBILITY OF SUCH DAMAGE.
*********************************************************************/

/* Author: Fernando Gonzalez fergonzaramos@yahoo.es  */

#include <iostream>
#include <memory>
#include <cstring>
#include "mavlink_encoder/checksum.h"
#include "mavlink_encoder/MavlinkEncoder.hpp"

#define MINPKGLEN 8

namespace mavlink_encoder
{
  /*
	 ***********************
   *   Public Methods    *
   ***********************
	 */

  MavlinkEncoder::MavlinkEncoder(int sys_id = 1, int comp_id = 0)
  : seq_n_(0)
  {
    this->sys_id_ = sys_id;
    this->comp_id_ = comp_id;
  }

  DigestType
	MavlinkEncoder::statusTextMsg(
    const std::string & msg, const status_text::StatusSeverity & severity)
  {
    DigestType out_digest;
    status_text::StatusTextEncoder *textEncoder;

    textEncoder = status_text::StatusTextEncoder::getInstance();

    // Get Payload
    DigestType payloadDigest;
    memset(payloadDigest.digest, 0, mavlink_encoder::MAXLENDIGEST);

    if (!textEncoder->composePayload(msg, severity, &payloadDigest))
		{
			free(textEncoder);
			throw StatusTextEncodeException();
		}
    
    // Get Header
    DigestType headerDigest;
    textEncoder->composeHeader(payloadDigest.len,
			seq_n_, sys_id_, comp_id_, &headerDigest);

		// Concatenate Header + Payload
    unsigned char digest[payloadDigest.len + headerDigest.len];
    memcpy(digest, headerDigest.digest, headerDigest.len);
    memcpy(&digest[headerDigest.len], payloadDigest.digest, payloadDigest.len);

    memcpy(out_digest.digest, digest, payloadDigest.len + headerDigest.len);
    out_digest.len = payloadDigest.len + headerDigest.len;

		// Get Checksum
    uint16_t check = checksum(out_digest, status_text::CrcExtra);

		memcpy(&out_digest.digest[out_digest.len], &check, 2);
    out_digest.len += 2;

    free(textEncoder);
    return out_digest;
  }

	bool
	MavlinkEncoder::decodePkg(
		const DigestType & digest, DigestMsgType * decoded_msg)
	{
		int header_len, payload_len;

		if (!msgIntegrityIsOk(digest))
			return false;

		// Get Header
		header_len = MINPKGLEN - 2;
		memcpy(decoded_msg->header.digest, digest.digest, header_len);
		decoded_msg->header.len = header_len;
		if (!headerIsOk(decoded_msg->header, digest))
			return false;

		// Get Payload
		memcpy(decoded_msg->payload.digest,
			&digest.digest[header_len], digest.digest[MINPKGLEN - 2]);
		decoded_msg->payload.len = digest.len - MINPKGLEN;

		// Get checksum
		memcpy(&decoded_msg->checksum, &digest.digest[digest.len - 2], 2);

		// Get message id
		decoded_msg->msg_type = static_cast<MsgsIds>(
			digest.digest[MINPKGLEN - 3]);

		return true;
	}

	bool
	MavlinkEncoder::msgIntegrityIsOk(const DigestType & full_pkg)
	{
		uint8_t crc_extra;

		uint8_t msg_id = full_pkg.digest[MINPKGLEN - 3];

		getCrcExtra(msg_id, &crc_extra);

		return (&crc_extra != nullptr) && crcIsOk(full_pkg, crc_extra);
	}

	void
	MavlinkEncoder::getCrcExtra(uint8_t id, uint8_t * crc_extra)
	{
		/* There are best ways to do it. When this library be capable of
		 * decode more types of packages, it will be improved for not
		 * having an enormous switch
		 */

		switch (id)
		{
		case MsgsIds::StatusText:
			*crc_extra = status_text::CrcExtra;
			break;

		default:
			crc_extra = nullptr;
		}
	}

	void
	MavlinkEncoder::increaseSeqN()
	{
		seq_n_++;
	}

	void
	MavlinkEncoder::setSeqN(int seq)
	{
		seq_n_ = seq;
	}

  /*
	 ***********************
   *   Private Methods   *
   ***********************
	 */

	bool
	MavlinkEncoder::headerIsOk(
		const DigestType & header, const DigestType & full_pkg)
	{
		// Decode all fields
		int header_len = MINPKGLEN - 2;

		return (header.digest[0] == 0xFE) &&
			(header.digest[1] == (full_pkg.len - MINPKGLEN)) &&
			(header.digest[2] >= 0) &&
			(header.digest[3] >= 1) &&
			(header.digest[4] >= 1);
	}

	bool
	MavlinkEncoder::crcIsOk(const DigestType & digest, uint8_t crc_extra)
	{
		// Calculate the checksum of the message excluding the crc
		DigestType msg_data;

		memcpy(&msg_data, digest.digest, digest.len - 2);
		msg_data.len = digest.len - 2;
		uint16_t msg_checksum = checksum(msg_data, crc_extra);

		// Separate checksum in two bytes
		unsigned char check[2];
		memcpy(check, &msg_checksum, 2);

		// Compare calculated checksum with the chacksum of the package
		return digest.digest[digest.len - 2] == check[0] &&
			digest.digest[digest.len - 1] == check[1];
	}

  uint16_t
	MavlinkEncoder::checksum(const DigestType & digest, uint8_t crc_extra)
  {
    // Ignore the first byte (magic byte -marker-)
  
    uint16_t checksum = crc_calculate(&digest.digest[1], digest.len - 1);
		crc_accumulate(crc_extra, &checksum);

    return checksum;
  }
}

File: MavlinkRawComm/src/mavlink_encoder/MavlinkEncoder.java
package mavlink_encoder;

import status_text.StatusTextEncoder;
import status_text.StatusTextPayloadEx;

import java.sql.Array;
import java.util.Arrays;

public class MavlinkEncoder {

    public static class MsgsIds {
        public static final int StatusText = 253;
    }

    public static class DigestMsg {
        public int msg_type;
        public byte[] header;
        public byte[] payload;
        public int checksum;
    }

    private static final int MinPkgLen = 8;
    private int seq_n = 0;
    private int sys_id, comp_id;
    public  MavlinkEncoder(int sys_id, int comp_id) {
        this.seq_n = 0;
        this.sys_id = sys_id;
        this.comp_id = comp_id;
        System.out.println("I'm Mavlink Encoder");
    }

    /**Returns a digest with full statustext message encoded following mavlink
     * packages specification
     *
     * @param msg Message to be sent (50 chars maximum)
     * @param severity Severity of the statustext message
     * @return Byte array that contains the full message encoded and ready
     *  to be sent to a Pixhawk/Autopilot
     * @throws StatusTextPayloadEx Thrown if the message is larger than 50 bytes
     */
    public byte[] statusTextMsg(String msg, int severity) throws StatusTextPayloadEx {
        byte[] payloadDigest, headerDigest, bodyPkg, checksum;
        StatusTextEncoder textEncoder = StatusTextEncoder.getInstance();

        // Compose Payload
        payloadDigest = textEncoder.composePayload(msg, severity);

        // Compose Header
        headerDigest = textEncoder.composeHeader(payloadDigest.length, seq_n, sys_id, comp_id);

        // Concatenate Header + Payload
        bodyPkg = concatenateDigests(headerDigest, payloadDigest);

        // Calculate checksum
        int check = checksum(Arrays.copyOfRange(bodyPkg, 1, bodyPkg.length), StatusTextEncoder.CRCEXTRA);

        checksum = new byte[2];
        checksum[0] = (byte) (check & 0xFF);
        checksum[1] = (byte) ((check >> 8) & 0xFF);

        return concatenateDigests(bodyPkg, checksum);
    }

    public DigestMsg decodePkg(byte[] digest) {
        DigestMsg digestMsg = null;
        if (msgIntegrityIsOk(digest)) {
            digestMsg = new DigestMsg();

            // Get header
            digestMsg.header = new byte[MinPkgLen - 2];
            for (int i = 0; i < MinPkgLen - 2; i++) {
                digestMsg.header[i] = digest[i];
            }

            // Get payload
            digestMsg.payload = new byte[digest.length - MinPkgLen];
            int j = 0;
            for (int i = MinPkgLen - 2; i < digest.length - 2; i++) {
                digestMsg.payload[j++] = digest[i];
            }

            // Get Checksum
            digestMsg.checksum = ((digest[digest.length - 1] & 0xff) << 8) | (digest[digest.length - 2] & 0xff);

            // Get message id
            digestMsg.msg_type = digest[MinPkgLen - 3] & 0xff;
        }

        return digestMsg;
    }

    /**
     * Returns if the codified mavlink message is correctly built. It checks the checksum
     * respecting to the package data
     *
     * @param full_pkg mavlink message
     * @return boolean indicating the integrity of the message
     */
    public boolean msgIntegrityIsOk(byte[] full_pkg) {
        byte crc_extra;
        int msg_id;

        msg_id = full_pkg[MinPkgLen - 3] & 0xFF;
        crc_extra = crcExtra(msg_id);

        return crcIsOk(full_pkg, crc_extra);
    }

    public byte crcExtra(int msg_id) {
        byte crc_extra;
        switch (msg_id) {
            case MsgsIds.StatusText:
                crc_extra = StatusTextEncoder.CRCEXTRA;
                break;
            default:
                crc_extra = 0x00 & 0xFF;
        }
        return crc_extra;
    }

    /**
     * Increase in one the sequence number. It will be encoded in the next package bytes calculated
     */
    public void increaseSeqN() {
        seq_n++;
    }

    /**Set a specific Sequence Number. It will be encoded in the next package bytes calculated
     *
     * @param seq_n Sequence number to set
     */
    public void setSeqN(int seq_n) {
        this.seq_n = seq_n;
    }

    /**
     *
     * @param digest Mavlink message to check its crc (checksum)
     * @param crc_extra crc extra corresponding to the type of message
     * @return true if crc is ok. Else, false is returned
     */
    private boolean crcIsOk(byte[] digest, byte crc_extra) {
        // Calculate the checksum of the message excluding the crc
        int msg_checksum = checksum(Arrays.copyOfRange(digest, 1, digest.length - 2), crc_extra);

        // Get checksum encoded in the package
        int pkg_checksum = ((digest[digest.length - 1] & 0xff) << 8) | (digest[digest.length - 2] & 0xff);

        return msg_checksum == pkg_checksum;
    }

    /**
     * Calculate a checksum from mavlink digest message
     *
     * @param buffer Digest to calculate the checksum
     * @param crc_extra crc extra corresponding to the type of message
     * @return the calculated checksum (integer)
     */
    private int checksum(byte[] buffer, byte crc_extra) {
        Checksum checksumComposer = new Checksum();
        checksumComposer.crcCalculate(buffer);
        checksumComposer.crcAccumulate(crc_extra);
        return checksumComposer.getCrc();
    }

    private byte[] concatenateDigests(byte[] digest1, byte[] digest2) {
        byte[] out = new byte[digest1.length + digest2.length];
        int i;

        for (i = 0; i < digest1.length; i++) {
            out[i] = digest1[i];
        }
        for (i = 0; i < digest2.length; i++) {
            out[digest1.length + i] = digest2[i];
        }

        return out;
    }
}
